# 프로미스 (Promise) 는 무엇일까요

- [자바스크립트의 비동기 프로그래밍은 무엇일까요](https://github.com/dbwjd5864/programmers_knowledge_storage/blob/main/Frontend/Javascript/%EC%9E%90%EB%B0%94%EC%8A%A4%ED%81%AC%EB%A6%BD%ED%8A%B8%EC%9D%98_%EB%B9%84%EB%8F%99%EA%B8%B0_%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D%EC%9D%80_%EB%AC%B4%EC%97%87%EC%9D%BC%EA%B9%8C%EC%9A%94.md) 에서 언급했던 것처럼 자바스크립트는 비동기 처리를 위한 하나의 패턴으로 콜백 함수를 사용한다
- 하지만 콜백 패턴은 콜백 헬을 발생시키고 콜백 헬은 아래와 같은 문제점들이 있다

    ```
    - 콜백 헬은 가독성을 나쁘게 한다
    - 비동기 처리 중 발생한 에러의 예외 처리가 곤란하다
    - 여러 개의 비동기 처리를 한 번에 처리하는 데도 한계가 있다
    ```

- 그리하여 비동기 처리를 위한 또 다른 패턴으로 도입된 것이 **프로미스 (Promise)** 이다

<br> 

## 콜백 패턴과 콜백 패턴의 단점


💡**프로미스 (Promise)** 를 본격적으로 알아보기 전에 콜백 헬이 구체적으로 무엇이고 단점이 무엇인지 알아보려고 한다

- 기대와는 다르게 동작하는 비동기 함수

```jsx
// GET 요청을 위한 비동기 함수
// 비동기 함수란, 함수 내부에 비동기로 동작하는 코드를 포함한 함수를 말한다

let post;

const getData = url => {
	const xhr = new XMLHttpRequest();
	xhr.open('GET', url);
	xhr.send();

	xhr.onload = () => {
		if(xhr.status === 200){
			// 1. 상위 스코프 변수에 할당
			post = JSON.parse(xhr.response);
			// 2. 서버의 응답을 반환
			return JSON.parse(xhr.response);
		}else{
			console.error(`${xhr.status} ${xhr.statusText}`);
		}
	};
};

// id가 1인 post를 취득
const response = getData('https://thisIsExample/posts/1');

console.log(response); // **undefined**
console.log(post); // **undefined**
```

```
아마 많은 사람들이 getData의 함수 호출 값으로 id가 1인 post를 취득하여 
콘솔에는 undefined가 아닌 실제 id가 1인 post 값이 찍히거나, 
상위 스코프인 post에 할당한 실제 값이 존재할 것이라 기대할 것이다

💡 하지만 getData 함수의 onload 이벤트 핸들러에서 서버의 응답 결과를 반환 (return) 하거나 
상위 스코프의 변수에 할당하면 기대한 대로 동작되지 않는다
```

- 비동기 함수 (**getData**) 를 호출 하면 함수 내부의 비동기로 동작하는 코드 (**onload 이벤트 핸들러**) 가 완료되지 않았다 해도 기다리지 않고 즉시 종료된다 ( 자바스크립트의 비동기 프로그래밍은 무엇일까요에서 다뤘던 **call stack 과 task queue의 관계**를 떠올려보자! )
- 즉, 비동기 함수인 getData 함수 내부의 **onload 이벤트 핸들러**는 **getData 함수가 종료된 이후에 실행된다**
    - getData 함수를 호출하면?
        1. GET 요청을 전송
        2. onload 이벤트 핸들러를 등록
        3. getData 함수에 명시적인 반환문이 없으므로 getData는 undefined를 반환하고 즉시 종료
1. 상위 스코프 변수에 할당
    - 비동기로 동작하는 **onload 이벤트 핸들러**는 언제나 **call stack** 에 담겨진 모든 실행 컨텍스트 (getData 함수, console.log 들도 여기에 포함돼있다) 가 종료된 후 실행된다

    ```jsx
    // 이는 곧 전역 변수인 post에 서버의 응답 결과가 할당되기 이전에 console.log가 실행됨을 의미한다
    // 💡 따라서 onload 이벤트 핸들러에서 상위 스코프 변수 (post) 에 서버의 응답 결과를 할당하기 이전에
    // conosle.log가 먼저 호출되어 **undefined를 출력한다**
    
    console.log(post); // **undefined**
    ```

2. 서버의 응답을 반환
    - onload 이벤트 핸들러를 getData 함수가 호출 할 수 있다면 이벤트 핸들러의 반환값을 getData 함수가 캐치하여 다시 반환할 수 있겠지만 onload 이벤트 핸들러는 getData 함수가 호출하지 않는다

    ```jsx
    // 💡 따라서 onload 이벤트 핸들러의 반환값은 캐치할 수 없고 reponse는 **undefined를 출력한다**
    const response = getData('https://thisIsExample/posts/1');
    
    console.log(response); // **undefined**
    ```
   
<br>

---

<br>

```
⭐️ 이처럼 비동기 함수는 비동기 처리 결과를 외부에 반환할 수 없고, 상위 스코프의 변수에 할당할 수 없다
```

- 따라서 비동기 함수의 처리 결과 (서버의 response 등) 에 대한 후속 처리는 비동기 함수 내부에서 수행해야 한다
- 이때 비동기 함수에 비동기 처리 결과에 대한 후속 처리를 수행하는 콜백 함수를 전달하는 것이 일반적이며 비동기 처리가 `성공하면 호출될 콜백 함수`와 비동기 처리가 `실패하면 호출될 콜백 함수`를 전달할 수 있다

```jsx
// GET 요청을 위한 비동기 함수
const getData = (url, successCallback, failureCallback) => {
	const xhr = new XMLHttpRequest();
	xhr.open('GET', url);
	xhr.send();

	xhr.onload = () => {
		if(xhr.status === 200){
			// 서버의 응답을 콜백 함수에 인수로 전달하면서 호출하여 응답에 대한 후속 처리를 한다
			successCallback(xhr.response);
		}else{
			// 에러 정보를 콜백 함수에 인수로 전달하면서 호출하여 에러 처리를 한다
			failureCallback(xhr.status);
		}
	};
};

// id가 1인 post를 취득
// 서버의 응답에 대한 후속 처리를 위한 콜백 함수를 비동기 함수인 getData에 전달해야 한다
getData('https://thisIsExample/posts/1', console.log, console.error);
```

- 콜백 함수를 통해 비동기 처리 결과에 대한 후속 처리를 수행하는 비동기 함수가 비동기 처리 결과를 가지고 또다시 비동기 함수를 호출해야 한다면 콜백 함수 호출이 중첩되어 복잡도가 높아지며 이를 **콜백 헬 (callback hell)** 이라 한다

```jsx
// 콜백 헬

const url = 'https://thisIsExample/posts/1';

// id가 1인 post userId를 취득
getData(url, ({ userId }) => {
	console.log(userId); // 1

	// post의 userId를 사용하여 user 정보를 취득
	getData(`${url}/users/${userId}`, userInfo => {
		console.log(userInfo) // { id: 1, name: "Yujeong", username: "Jade", ... }
	});
});
```

- 위 예제를 보면 GET 요청을 통해 서버로부터 응답 (id가 1인 post)을 취득하고, 이 데이터를 사용하여 또다시 GET 요청을 한다

<br> 

### 콜백 헬의 단점

---

1. 콜백 헬은 **가독성을 나쁘게 한다**

```jsx
// 또 다른 콜백 헬 예시
getData('/step1', a => {
	getData(`/step2/${a}`, b => {
		getData(`/step2/${b}`, c => {
			getData(`/step2/${c}`, d => {
				console.log(d);
			})
		})
	})
})
```

1. 또 다른 콜백 패턴의 문제점은 **에러 처리가 곤란**하다는 것이다

```jsx
try{
	setTimeout(() => {
		throw new Error('ERROR!');
	}, 1000);
}catch (e){
	// 에러를 캐치하지 못한다
	console.error('캐치한 에러', e);
}
```

- try 코드 블록 내에서 호출한 setTimeout 함수는 1초 후에 콜백 함수가 실행되도록 타이머를 설정하고, 이후 콜백 함수는 에러를 발생시킨다.
- 하지만 이 에러는 catch 코드 블록에서 캐치되지 않는다
    - 비동기 함수인 setTimeout 는 호출되면 setTimeout 함수의 실행 컨텍스트가 생성되어 콜 스택에 푸쉬되며 실행된다
    - setTimeout 은 비동기 함수이므로 콜백 함수가 호출되는 것을 기다리지 않고 즉시 종료되어 콜 스택에서 제거된다
    - 이후 타이머가 만료되면 setTimeout 함수의 콜백 함수는 태스크 큐로 푸시되고 콜 스택이 비어졌을 때 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다

    ```
    1. setTimeout 함수의 콜백 함수가 실행됐을 때는 setTimeout 함수는 이미 콜 스택에서 제거된 상태다
    	- 이것은 setTimeout 함수의 콜백 함수를 호출한 것이 setTimeout 함수가 아님을 의미한다
    	- setTimeout 함수의 콜백 함수를 호출한 것이 setTimeout 함수라면 콜 스택에서 현재 실행 중인
    	실행 컨텍스트가 콜백 함수의 실행 컨텍스트 일 때 현재 실행 중인 콜백 함수의 실행 컨텍스트의 하위 실행
    	컨텍스트가 setTimeout 이어야 한다
    
    2. **에러는 호출자 (caller) 방향으로 전파된다. 즉, 콜 스택의 아래 방향으로 전파 된다.
    (현재 실행 중인 실행 컨텍스트가 푸시되기 직전에 푸시된 실행 컨텍스트 방향)**
    
    3. 하지만 setTimeout 함수의 콜백 함수를 호출한 것은 setTimeout 함수가 아니다
    
    4. 따라서 setTimeout 함수의 콜백 함수가 발생시킨 에러는 catch 블록에서 캐치되지 않는다
    (catch 블럭을 담고 있던 실행 컨텍스트는 setTimeout의 콜백 함수가 콜 스택으로 오기전에 이미 제거됐다)
    ```


## 프로미스 (Promise)

---

<aside>
💡 이러한 콜백 패턴의 단점을 극복하기 위해 프로미스 ( Promise ) 가 도입되었다
</aside>

<br> 

### 프로미스의 생성

- Promise 생성자 함수를 new 연산자와 함께 호출하면 프로미스  (Promise 객체) 를 생성한다
- Promise 생성자 함수는 비동기 처리를 수행할 **resolve** 와 **reject** 콜백 함수를 인수로 전달받는다

```jsx
const promise = new Promise((resolve, reject) => {
	// Promise 함수의 콜백 함수 내부에서 비동기 처리를 수행한다
	if(/* 비동기 처리 성공을 나타내는 코드 */){
		resolve('success result');
	}
	else{ // 비동기 처리 실패
		reject('failure reason');
	}
})

/* --------------------------------------------------- */

// 앞에서 사용했던 비동기 함수 getData에 프로미스를 활용해보자
const promiseGetData = url => {
	return new Promise((resolve, reject) => {
		const xhr = new XMLHttpRequest();
		xhr.open('GET', url);
		xhr.send();
	
		xhr.onload = () => {
			if(xhr.status === 200){
				// 성공적으로 응답을 전달받으면 resolve 함수를 호출한다
				resolve(JSON.parse(xhr.response));
			}else{
				// 에러 처리를 위해 reject 함수를 호출한다
				reject(new Error(xhr.status));
			}
		};
	}
};

// promiseGetData 함수는 프로미스를 반환한다
promiseGetData('https://thisIsExample/posts/1');
```

- 비동기 함수인 promiseGetData 는 함수 내부에서 프로미스를 생성하고 반환한다
- 비동기 처리는 Promise 생성자 함수가 인수로 전달받은 콜백 함수 내부에서 수행한다
- 비동기 처리는 현재 어떻게 진행되고 있는지를 나타내는 상태 정보를 갖는다

| 상태 정보 | 의미 | 상태 변경 조건 |
| --- | --- | --- |
| pending | 비동기 처리가 아직 수행되지 않은 상태 | 프로미스가 생성된 직후 기본 상태 |
| fulfilled | 비동기 처리가 수행된 상태 (성공) | resolve 함수 호출 |
| rejected | 비동기 처리가 수행된 상태 (실패) | reject 함수 호출 |
- 생성된 직후의 프로미스는 기본적으로 pending 상태이며 이후 비동기 처리 결과에 따라 프로미스의 상태가 변경된다
    - 비동기 처리 성공: resolve 함수 호출, 프로미스 상태를 fulfilled로 변경
    - 비동기 처리 실패: reject 함수 호출, 프로미스 상태를 rejected로 변경
    - fulfilled 혹은 rejected 상태를 **settled** 상태라고 하며 비동기 처리가 수행된 상태를 말한다
- 즉, **프로미스는 비동기 처리 상태와 처리 결과를 관리하는 객체이다**

<br> 

## 프로미스의 후속 처리 메서드


- 프로미스의 비동기 처리 상태가 변화 (pending ⇒ fulfilled, pending ⇒ rejected) 하면 이에 따른 후속 처리를 해야한다
- 이러한 후속 처리를 위해 프로미스는 후속 메서드인 then, catch, finally를 제공한다
- 모든 `후속 처리 메서드는 프로미스를 반환`하며, 비동기로 동작한다

<br> 

### 1. then

```
then 메서드는 두개의 콜백 함수를 인수로 전달받는다

1. 첫번째 콜백 함수: 프로미스가 fulfilled 상태 (resolve 함수가 호출된 상태)가 되면 호출된다
								이때 콜백 함수는 프로미스의 비동기 처리 결과를 인수로 전달받는다 
2. 두번째 콜백 함수: 프로미스가 rejected 상태 (reject 함수가 호출된 상태)가 되면 호출된다
								이때 콜백 함수는 프로미스의 에러를 인수로 전달받는다 
```

- then 메서드는 언제나 프로미스를 반환한다
    - then 메서드의 콜백 함수가 프로미스를 반환하면 그 프로미스를 그대로 반환한다
    - 콜백 함수가 프로미스가 아닌 값을 반환하면 그 값을 암묵적으로 resolve 또는 reject 하여 프로미스를 생성해 반환한다

```jsx
// fulfilled
const promiseFulfilled = new Promise(resolve => resolve('fulfilled'));
promiseFulfilled.then(result => console.log(result), error => console.error(error)); // fulfilled

// rejected
const promiseRejected = new Promise((_, reject) => reject(new Error('rejected')));
promiseRejected.then(result => console.log(result), error => console.error(error)); // Error: rejected
```

<br> 

### 2. catch

```
catch 메서드는 한개의 콜백 함수를 인수로 전달받는다

1. 콜백 함수: 프로미스가 rejected 상태인 경우만 호출된다
```

- catch 메서드는 then의 두번째 콜백 함수와 동일하게 동작한다. 따라서 then 메서드와 동일하게 언제나 프로미스를 반환한다

```jsx
// rejected
const promiseRejected = new Promise((_, reject) => reject(new Error('rejected')));
promiseRejected.catch(error => console.error(error)); // Error: rejected
```

<br> 

### 3. finally

```
finally 메서드는 한개의 콜백 함수를 인수로 전달받는다

1. 콜백 함수: 프로미스의 성공 (fulfilled) 또는 실패 (rejected)와 상관없이 무조건 한 번 호출된다
```

- then / catch 메서드와 마찬가지로 언제나 프로미스를 반환한다

<br> 

## 프로미스의 에러 처리

- 프로미스는 콜백 패턴과 다르게 에러를 문제없이 처리할 수 있다
- 프로미스를 활용한 비동기 처리에서 발생한 에러는 **then 메서드의 두 번째 콜백 함수**와 **catch**를 사용해 처리 할 수 있다
    - 단, then 메서드의 두 번째 콜백 함수는 첫 번째 콜백 함수에서 발생한 에러를 캐치하지 못하고 코드가 복잡해져서 가독성이 떨어지는 단점이 있다
    - catch 메서드를 모든 then 메서드를 호출한 이후에 호출하면 비동기 처리에서 발생한 에러 (rejected 상태) 뿐만 아니라 then 메서드 내부에서 발생한 에러까지 모두 잡을 수 있다
    - 따라서 에러 처리는 then 메서드가 아닌 catch 메서드에서 하는 것을 권장한다

<br> 

## 프로미스 체이닝

- 프로미스는 then, catch, finally 후속 처리 메서드를 통해 콜백 헬을 해결한다

```jsx
const promiseGetData = url => {
	return new Promise((resolve, reject) => {
		const xhr = new XMLHttpRequest();
		xhr.open('GET', url);
		xhr.send();
	
		xhr.onload = () => {
			if(xhr.status === 200){
			
				resolve(JSON.parse(xhr.response));
			}else{
				
				reject(new Error(xhr.status));
			}
		};
	}
};

const url = 'https://thisIsExample'
promiseGetData(`${url}/posts/1`)
.then(({userId}) => {
		return promiseGetData(`${url}/users/${userId}`)
})
.then(userInfo => {
		console.log(userInfo);
})
.catch(err => {
		console.error(err);
});
```

- then ⇒ then ⇒ catch 순서대로 후속 처리 메서드를 호출한다. then, catch, finally는 언제나 프로미스를 반환하므로 연속적으로 호출할 수 있고 이를 **프로미스 체이닝**이라고 한다
- 프로미스는 프로미스 체이닝을 통해 비동기 처리 결과를 전달받아 후속 처리를 하므로 콜백 헬이 발생하지 않는다
- 하지만! 프로미스도 콜백 패턴을 사용하므로 콜백 함수는 사용하고 있다. 콜백 함수는 가독성이 좋지 않고 이를 해결하기 위해 **async/await**이 도입되었다. **async/await**을 활용하면 마치 동기 처리처럼 프로미스가 처리 결과를 반환하도록 구현할 수 있다 (이에 대해서는 후에 ‘async/await은 무엇일까요’에서 더 자세하게 다뤄보려 한다)

<br> 

## 프로미스의 정적 메서드

- Promise는 주로 생성자 함수로 사용되지만 함수도 객체이므로 메서드를 가진다

<br> 

### 1. Promise.resolve / Promise.reject

- Promise.resolve 와 Promise.reject 메서드는 이미 존재하는 값을 래핑하여 프로미스를 생성한다

```jsx
// 배열을 resolve 하는 프로미스를 생성
const promiseResolved = Promise.resolve([1,2,3]);
promiseResolved.then(result => console.log(result)); // [1,2,3]

// 에러 객체를 reject 하는 프로미스를 생성
const promiseRejected = Promise.reject(new Error('Error!'));
promiseRejected.catch(err => console.error(err)); // Error: Error!
```

- 위 예제는 아래와 동일하게 동작한다

```jsx
const promiseResolved = new Promise(resolve => resolve([1,2,3]));
promiseResolved.then(result => console.log(result)); // [1,2,3]

const promiseRejected = new Promise((_,reject) => reject(new Error('Error!')));
promiseRejected.catch(err => console.error(err)); // Error: Error!
```

<br> 

### 2. Promise.all

- Promise.all 메서드는 여러 개의 비동기 처리를 모두 병렬 처리할 때 사용한다

```jsx
const promise1 = () => new Promise(resolve => setTimeout(() => resolve(1), 3000));
const promise2 = () => new Promise(resolve => setTimeout(() => resolve(2), 2000));
const promise3 = () => new Promise(resolve => setTimeout(() => resolve(3), 1000));

// 세개의 비동기를 순차적으로 처리
const res = [];
promise1
.then(data => {
	res.push(data);
	return promsie2();
})
.then(data => {
	res.push(data);
	return promsie3();
})
.then(data => {
	res.push(data);
	console.log(res); // [1, 2, 3] => 약 6초 소요
})
.catch(err => {
	console.error(err);
})
```

- 위 예제는 세개의 비동기 처리를 순차적으로 처리한다. 즉 앞선 비동기 처리가 완료되면 다음 비동기 처리를 수행한다
    - 첫번째 비동기 처리 3초
    - 두번째 비동기 처리 2초
    - 마지막 비동기 처리 1초
    - 총 6초 이상이 소요된다
    - 하지만 위와 같은 예시는 앞선 비동기 처리 결과를 다음 비동기 처리에서 사용하지 않으므로 순차적으로 처리할 필요가 없다
    - 이를 Promise.all를 이용해 병렬 처리하도록 바꿔보자

    ```jsx
    const promise1 = () => new Promise(resolve => setTimeout(() => resolve(1), 3000));
    const promise2 = () => new Promise(resolve => setTimeout(() => resolve(2), 2000));
    const promise3 = () => new Promise(resolve => setTimeout(() => resolve(3), 1000));
    
    // 세개의 비동기를 병렬로 처리
    Promise.all([promise1(), promise2(), promise3()])
    .then(result => {
    	console.log(result); // [1, 2, 3] => 약 3초 소요
    })
    .catch(err => {
    	console.error(err);
    })
    ```

- Promise.all 메서드는 인수로 전달받은 배열의 모든 프로미스가 모두 fulfilled 상태가 되면 종료된다
- 따라서 Promise.all 메서드가 종료하는데 걸리는 시간은 가장 늦게 fulfilled 상태가 되는 프로미스 처리 시간보다 조금 더 길다
- 위에 예시 같은 경우 첫 번째 프로미스가 가장 늦게 fulfilled 상태가 되어도 Promise.all 메서드는 첫 번째 프로미스가 resolve 한 결과부터 차례대로 배열에 저장해 그 배열을 resolve 하는 새로운 프로미스를 반환한다
- 즉, **처리 순서가 보장된다**
- Promise.all 배열의 프로미스가 하나라도 rejected 상태가 되면  나머지 프로미스가 fulfilled 상태가 되는 것을 기다리지 않고 즉시 종료한다

<br> 

### 3. Promise.race

- Promise.race 는 Promise.all 메서드와 다르게 가장 먼저 fulfilled 상태가 된 프로미스의 처리 결과를 resolve 하는 새로운 프로미스를 반환한다

```jsx
const promise1 = () => new Promise(resolve => setTimeout(() => resolve(1), 3000));
const promise2 = () => new Promise(resolve => setTimeout(() => resolve(2), 2000));
const promise3 = () => new Promise(resolve => setTimeout(() => resolve(3), 1000));

Promise.race([promise1(), promise2(), promise3()])
.then(result => {
	console.log(result); // 3
})
.catch(err => {
	console.error(err);
})
```

- Promise.race 배열의 프로미스가 하나라도 rejected 상태가 되면 에러를 reject 하는 새로운 프로미스를 즉시 반환한다

```
❗️ 사실상 실무에서 Promise.race는 많이 사용할 일이 없어 검색을 조금 더 해본 결과 서버로부터 일정시간 응답이 없을경우 수동으로 오류 처리를 하기 위해 사용되는 경우들은 있다는걸 알게 됐다

예를들어 db에서 특정 정보를 불러와서 화면에 랜더링하는데 오랜 시간이 걸린다면 로딩처리가 잘 되어있다 해도 사용자들은 화면을 볼 수 없게되고 불편함을 느끼게된다.
그렇기 때문에 특정 시간 이상 렌더링이 걸리게되는 데이터 호출은 timeout 프로미스와 db에서 데이터를 불러오는 로직을 Promise.race로 경합을 시키고
먼저 완료되는 프로미스를 반환하게 하여 특정 시간 이상 걸리게 됐을 때 다른 처리를 하게 만든다는지 정도로 사용할 수는 있다

아래는 이에 해당하는 예시이다
```

```javascript
const timeout = (s) => {
	return new Promise(resolve => setTimeout(() => resolve(s), s * 1000));
}

Promise.race([ 
	timeout(10).then((seconds) => {
		throw new Error(`${seconds} 초 이내에 완료되지 않음`)
	}),
]).catch(err => {
// 데이터 호출에 있어 특정 시간 이상 걸렸을 때 처리 코드
  console.log(err);  
})
```

<br> 

### 4. Promise.allSettled

- Promise.allSettled 메서드는 전달받은 프로미스가 **모두** settled 상태 (fulfilled 또는 rejected 상태)가 되면 처리 결과를 배열로 반환한다
- fulfilled 또는 rejected 상태와는 상관없이 모든 처리 결과가 담겨 있다

```jsx
const promiseFulfilled = () => new Promise(resolve => setTimeout(() => resolve(1), 3000));
const promiseRejected = () => new Promise((_, reject) => reject(new Error('Error')));

Promise.allSettled([promiseFulfilled(), promiseRejected()])
.then(result => {
	console.log(result); // 3
})
.catch(err => {
	console.error(err);
})
```

```jsx
[
	//프로미스가 fulfilled 상태인 경우 - value
    {
        "status": "**fulfilled**",
        "**value**": 1
    },
//프로미스가 rejected 상태인 경우 - reason
    {
        "status": "**rejected**",
        "**reason**": Error: Error at <anonymous>:3:58
    }
]
```

<br> 

## 마이크로태스크 큐

```jsx
// 어떤 순서대로 출력될까?
setTimeout(() => console.log(1), 0);

Promise.resolve()
.then(() => console.log(2))
.then(() => console.log(3));  
```

- setTimeout 의 콜백 함수 및 Promise 의 후속 처리 메서드는 비동기 동작이므로 순서대로 1 ⇒ 2 ⇒ 3 이 출력될 것이라 생각할 수 있다. 하지만 출력은 **2 ⇒ 3 ⇒ 1** 순서대로 진행된다
    - 그 이유는? **프로미스의 후속 처리 메서드의 콜백 함수는 태스크 큐가 아니라 마이크로태스크 큐 (microtask queue) 에 저장**되기 때문이다
- 마이크로태스크 큐는 태스크 큐와는 별도의 큐이다
    - 마이크로태스크 큐에는 프로미스의 후속 처리 메서드의 콜백 함수가 일시 저장된다
    - 그 외의 비동기 함수의 콜백 함수나 이벤트 핸들러는 태스크 큐에 일시 저장된다
- **마이크로태스크 큐는 태스크 큐보다 우선순위가 높다**

<aside>
💡 즉, 이벤트 루프는 콜 스택이 비면 마이크로태스크 큐에서 대기하고 있는 함수를 먼저 가져와 실행한다
이후, 마이크로태스크 큐가 비면 태스크 큐에서 대기하고 있는 함수를 가져와 실행한다

</aside>

<br>

---

##### Reference   

- [모던 자바스크립트 Deep Dive](https://poiemaweb.com/es6-promise)  
